routesrv/routesrv.go

```diff
@@ -0,0 +1,154 @@
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/dataclients/kubernetes"
+	"github.com/zalando/skipper/tracing"
+)
+
+// RouteServer is used to serve eskip-formatted routes,
+// that originate from the polled data source.
+type RouteServer struct {
+	server *http.Server
+	poller *poller
+	wg     *sync.WaitGroup
+}
+
+// New returns an initialized route server according to the passed options.
+// This call does not start data source updates automatically. Kept routes
+// will stay in an uninitialized state, till StartUpdates is called and
+// in effect data source is queried and routes initialized/updated.
+func New(opts Options) (*RouteServer, error) {
+	rs := &RouteServer{}
+
+	opentracingOpts := opts.OpenTracing
+	if len(opentracingOpts) == 0 {
+		opentracingOpts = []string{"noop"}
+	}
+	tracer, err := tracing.InitTracer(opentracingOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &eskipBytes{tracer: tracer}
+	bs := &eskipBytesStatus{b: b}
+	handler := http.NewServeMux()
+	handler.Handle("/health", bs)
+	handler.Handle("/routes", b)
+	handler.Handle("/metrics", promhttp.Handler())
+	rs.server = &http.Server{Addr: opts.Address, Handler: handler}
+
+	dataclient, err := kubernetes.New(kubernetes.Options{
+		KubernetesInCluster:               opts.KubernetesInCluster,
+		KubernetesURL:                     opts.KubernetesURL,
+		ProvideHealthcheck:                opts.KubernetesHealthcheck,
+		ProvideHTTPSRedirect:              opts.KubernetesHTTPSRedirect,
+		HTTPSRedirectCode:                 opts.KubernetesHTTPSRedirectCode,
+		IngressClass:                      opts.KubernetesIngressClass,
+		RouteGroupClass:                   opts.KubernetesRouteGroupClass,
+		ReverseSourcePredicate:            opts.ReverseSourcePredicate,
+		WhitelistedHealthCheckCIDR:        opts.WhitelistedHealthCheckCIDR,
+		PathMode:                          opts.KubernetesPathMode,
+		KubernetesNamespace:               opts.KubernetesNamespace,
+		KubernetesEastWestRangeDomains:    opts.KubernetesEastWestRangeDomains,
+		KubernetesEastWestRangePredicates: opts.KubernetesEastWestRangePredicates,
+		DefaultFiltersDir:                 opts.DefaultFiltersDir,
+		BackendNameTracingTag:             opts.OpenTracingBackendNameTag,
+		OnlyAllowedExternalNames:          opts.KubernetesOnlyAllowedExternalNames,
+		AllowedExternalNames:              opts.KubernetesAllowedExternalNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+	rs.poller = &poller{
+		client:  dataclient,
+		timeout: opts.SourcePollTimeout,
+		b:       b,
+		quit:    make(chan struct{}),
+		tracer:  tracer,
+	}
+
+	rs.wg = &sync.WaitGroup{}
+
+	return rs, nil
+}
+
+// StartUpdates starts the data source polling process.
+func (rs *RouteServer) StartUpdates() {
+	rs.wg.Add(1)
+	go rs.poller.poll(rs.wg)
+}
+
+// StopUpdates stop the data source polling process.
+func (rs *RouteServer) StopUpdates() {
+	rs.poller.quit <- struct{}{}
+}
```

Comment:
`wg` stayed here only for convenience when `Run` call is used: https://github.com/zalando/skipper/pull/1848/files#diff-903c01df2b8ad59771be10eabada676a9161af143e6ce387f9abe60070af5eb3R128 in order to wait for all the async activities with a single barrier. `StartUpdates` is not meant to be called multiple times, but you're right that now no one prevents doing it and neighboring `WaitGroup` can suggest so.

Reply:
> it should close the channel and wait on the waitgroup for all threads to complete

it can not since we reuse wg for waiting for shutdown completion. IMO we should not reuse but rather have a shutdown channel like 
https://github.com/zalando/skipper/blob/63b27681807f962a15372c41b43bb0ceea575068/skipper.go#L1080

(name `idleConnsCH` is a bit misleading though)

Reply:
> StartUpdates is not meant to be called multiple times

Then there is no need to tick wg here and pass it to the poller at all, the quit channel should be enough to control the poller

Reply:
I think it would be beneficial if we divide the discussion to separate topics and conquer with a suited solution:
1. Do we want / is it beneficial to run multiple polling goroutines?
2. Do we want to enable them to be stopped and started as many times as needed or just start once and be able to stop once?

IMHO Ad 1. for the time being this is not needed, 2. I do not think this is needed as well.

Reply:
> Then there is no need to tick wg here and pass it to the poller at all, the quit channel should be enough to control the poller

You're right. For the current shape of the code yes, because there is no real action taken when `quit` "message" is received. The other signal would only be needed if we really wanted to wait for the poller to finish a potential cleanup / actions after the quit signal had been received. My thinking was rather for correctness to make sure the goroutine is really over and guard against potential bugs as well when some handling is added for handling quit signal in the poller. It might have been a premature optimization. I am not really attached to this solution, I rather treat is a matter of personal preference / style.

Reply:
@szuecs @aryszka what do you think?